commit 216df8ac250e2fe705c90f07c2a5e1811e7011e9
Author: jim-p <jimp@netgate.com>
Date:   Mon Mar 25 14:15:02 2024 -0400

    Disallow hostnames in Kea NTP. Fixes #14991
    
    * If they are in the config.xml data already, do not write them into
      the Kea configuration.
    * Do not allow the user to enter them in the GUI

diff --git a/src/etc/inc/services.inc b/src/etc/inc/services.inc
index 8387df746a..d36307eff4 100644
--- a/src/etc/inc/services.inc
+++ b/src/etc/inc/services.inc
@@ -1420,7 +1420,7 @@ function services_kea4_configure()
 			if (is_array($poolconf['ntpserver']) && $poolconf['ntpserver'][0] && ($poolconf['ntpserver'][0] != $dhcpifconf['ntpserver'][0])) {
 				$keapool['option-data'][] = [
 					'name' => 'ntp-servers',
-					'data' => implode(', ', $poolconf['ntpserver'])
+					'data' => implode(', ', array_filter($poolconf['ntpserver'], 'is_ipaddrv4'))
 				];
 			}
 
@@ -1611,7 +1611,7 @@ function services_kea4_configure()
 		if (is_array($dhcpifconf['ntpserver']) && $dhcpifconf['ntpserver'][0]) {
 			$keasubnet['option-data'][] = [
 				'name' => 'ntp-servers',
-				'data' => implode(', ', $dhcpifconf['ntpserver'])
+				'data' => implode(', ', array_filter($dhcpifconf['ntpserver'], 'is_ipaddrv4'))
 			];
 		}
 
@@ -1850,7 +1850,7 @@ function services_kea4_configure()
 				if (is_array($sm['ntpserver']) && $sm['ntpserver'][0] && ($sm['ntpserver'][0] != $dhcpifconf['ntpserver'][0])) {
 					$keares['option-data'][] = [
 						'name' => 'ntp-servers',
-						'data' => implode(', ', $sm['ntpserver'])
+						'data' => implode(', ', array_filter($sm['ntpserver'], 'is_ipaddrv4'))
 					];
 				}
 
diff --git a/src/usr/local/www/services_dhcp.php b/src/usr/local/www/services_dhcp.php
index 039cd35984..72d7b2c3b0 100644
--- a/src/usr/local/www/services_dhcp.php
+++ b/src/usr/local/www/services_dhcp.php
@@ -450,11 +450,15 @@ if (isset($_POST['save'])) {
 		$input_errors[] = gettext("If a mac deny list is specified, it must contain only valid partial MAC addresses.");
 	}
 
-	if (($_POST['ntp1'] && (!is_ipaddrv4($_POST['ntp1']) && !is_hostname($_POST['ntp1']))) ||
-	    ($_POST['ntp2'] && (!is_ipaddrv4($_POST['ntp2']) && !is_hostname($_POST['ntp2']))) ||
-	    ($_POST['ntp3'] && (!is_ipaddrv4($_POST['ntp3']) && !is_hostname($_POST['ntp3']))) ||
-	($_POST['ntp4'] && (!is_ipaddrv4($_POST['ntp4']) && !is_hostname($_POST['ntp4'])))) {
-		$input_errors[] = gettext("A valid IP address or hostname must be specified for the NTP servers.");
+	if (($_POST['ntp1'] && !(is_ipaddrv4($_POST['ntp1']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp1'])))) ||
+	    ($_POST['ntp2'] && !(is_ipaddrv4($_POST['ntp2']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp2'])))) ||
+	    ($_POST['ntp3'] && !(is_ipaddrv4($_POST['ntp3']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp3'])))) ||
+	    ($_POST['ntp4'] && !(is_ipaddrv4($_POST['ntp4']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp4']))))) {
+		if (dhcp_is_backend('isc')) {
+			$input_errors[] = gettext("A valid IP address or hostname must be specified for the NTP servers.");
+		} else {
+			$input_errors[] = gettext("A valid IP address must be specified for the NTP servers.");
+		}
 	}
 	if ($_POST['domain'] && (!is_domain($_POST['domain'], false, false))) {
 		$input_errors[] = gettext("A valid domain name must be specified for the DNS domain.");
diff --git a/src/usr/local/www/services_dhcp_edit.php b/src/usr/local/www/services_dhcp_edit.php
index 029ee7aade..9cac8fc3a2 100644
--- a/src/usr/local/www/services_dhcp_edit.php
+++ b/src/usr/local/www/services_dhcp_edit.php
@@ -317,10 +317,15 @@ if ($_POST['save']) {
 		}
 	}
 
-	if (($_POST['ntp1'] && (!is_ipaddrv4($_POST['ntp1']) && !is_hostname($_POST['ntp1']))) ||
-	    ($_POST['ntp2'] && (!is_ipaddrv4($_POST['ntp2']) && !is_hostname($_POST['ntp2']))) ||
-	    ($_POST['ntp3'] && (!is_ipaddrv4($_POST['ntp3']) && !is_hostname($_POST['ntp3'])))) {
-		$input_errors[] = gettext("A valid IP address or hostname must be specified for the primary/secondary NTP servers.");
+	if (($_POST['ntp1'] && !(is_ipaddrv4($_POST['ntp1']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp1'])))) ||
+	    ($_POST['ntp2'] && !(is_ipaddrv4($_POST['ntp2']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp2'])))) ||
+	    ($_POST['ntp3'] && !(is_ipaddrv4($_POST['ntp3']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp3'])))) ||
+	    ($_POST['ntp4'] && !(is_ipaddrv4($_POST['ntp4']) || (dhcp_is_backend('isc') && is_hostname($_POST['ntp4']))))) {
+		if (dhcp_is_backend('isc')) {
+			$input_errors[] = gettext("A valid IP address or hostname must be specified for the NTP servers.");
+		} else {
+			$input_errors[] = gettext("A valid IP address must be specified for the NTP servers.");
+		}
 	}
 	if ($_POST['domain'] && (!is_domain($_POST['domain'], false, false))) {
 		$input_errors[] = gettext("A valid domain name must be specified for the DNS domain.");
